lib/plugins/aws/invokeLocal/index.js

```diff
@@ -665,99 +635,95 @@ class AwsInvokeLocal {
-        fse
-          .stat(executablePath)
-          .then(() => this.callJavaBridge(artifactPath, className, handlerName, input))
-          .then(resolve, () => {
-            const mvn = spawn('mvn', ['package', '-f', path.join(javaBridgePath, 'pom.xml')], {
-              shell: true,
-            });
+    const javaBridgePath = await this.resolveRuntimeWrapperPath('java');
+    const executablePath = path.join(javaBridgePath, 'target');
 
-            this.serverless.cli.log(
-              'Building Java bridge, first invocation might take a bit longer.'
-            );
+    await fse.stat(executablePath);
 
-            mvn.stderr.on('data', buf =>
-              this.serverless.cli.consoleLog(`mvn(stderr) - ${buf.toString()}`)
-            );
-            const chunk = [];
-            if (process.env.SLS_DEBUG) {
-              mvn.stdout.on('data', buf => chunk.push(buf));
-            }
+    try {
+      return await this.callJavaBridge(artifactPath, className, handlerName, input);
+    } catch (err) {
+      return new Promise((resolve, reject) => {
+        const mvn = spawn('mvn', ['package', '-f', path.join(javaBridgePath, 'pom.xml')], {
+          shell: true,
+        });
 
-            let isRejected = false;
-            mvn.on('error', error => {
-              if (!isRejected) {
-                isRejected = true;
-                reject(error);
-              }
-            });
+        this.serverless.cli.log('Building Java bridge, first invocation might take a bit longer.');
 
-            mvn.on('exit', (code, signal) => {
-              if (code === 0) {
-                this.callJavaBridge(artifactPath, className, handlerName, input).then(resolve);
-              } else if (!isRejected) {
-                if (process.env.SLS_DEBUG) {
-                  chunk
-                    .map(elem => elem.toString())
-                    .join('')
-                    .split(/\n/)
-                    .forEach(line => {
-                      this.serverless.cli.consoleLog(`mvn(stdout) - ${line}`);
-                    });
-                }
-                isRejected = true;
-                reject(
-                  new Error(`Failed to build the Java bridge. exit code=${code} signal=${signal}`)
-                );
-              }
-            });
+        mvn.stderr.on('data', buf =>
+          this.serverless.cli.consoleLog(`mvn(stderr) - ${buf.toString()}`)
+        );
+        const chunk = [];
+        if (process.env.SLS_DEBUG) {
+          mvn.stdout.on('data', buf => chunk.push(buf));
+        }
 
-            process.nextTick(() => {
-              if (isRejected) return; // Runtime not available
-              mvn.stdin.end();
-            });
-          })
-      );
-    });
+        let isRejected = false;
+        mvn.on('error', error => {
+          if (!isRejected) {
+            isRejected = true;
+            reject(error);
+          }
+        });
+
+        mvn.on('exit', async (code, signal) => {
```

Comment:
This handler should not be an _async_ function as it's result is not processed (technically that creates an orphaned promise)

Reply:
I disagree on this one.

And an orphaned promise is a promise that's never resolved so stays there taking up memory and event loop cycles.

This is just a promise that we don't care about the result.
It will be invoked by the internals of `mvn.on` and it will be garbage collected whenever it resolved and `mvn` is gc'd to.

The previous code was also using a promise which result was not processed (`callJavaBridge`)
```js
            mvn.on('exit', (code, signal) => {
              if (code === 0) {
                this.callJavaBridge(artifactPath, className, handlerName, input).then(resolve);	        );
              } else if (!isRejected) {
```

We have plenty of promises that we don't process the results, like whenever we do `await fs.lstat(file)` to `catch` if the file exists.

Does that make sense?
Or would you still like me to make it sync?

Reply:
>And an orphaned promise is a promise that's never resolved

It is resolved, and if rejected, it'll end as unhandled rejection

The right way to do that is as follows:

```javascript
mvn.on('exit', (code, signal) => {
  if (code === 0) {
    resolve(this.callJavaBridge(artifactPath, className, handlerName, input)));
  } ...
});
```